errors_test.go

```diff
@@ -44,6 +45,119 @@ func TestInvalidArgument(t *testing.T) {
 	}
 }
 
+func TestErrorEquivalence(t *testing.T) {
+	var e1 error = ErrAborted
+	var e2 error = ErrUnknown
+	if e1 == e2 {
+		t.Fatal("should not equal the same error")
+	}
+	if errors.Is(e1, e2) {
+		t.Fatal("errors.Is should not return true")
+	}
+
+	var e3 error = errAborted{}
+	if e1 != e3 {
+		t.Fatal("new instance should be equivalent")
+	}
+	if !errors.Is(e1, e3) {
+		t.Fatal("errors.Is should be true")
+	}
+	if !errors.Is(e3, e1) {
+		t.Fatal("errors.Is should be true")
+	}
+	var aborted errAborted
+	if !errors.As(e1, &aborted) {
+		t.Fatal("errors.As should be true")
+	}
+
+	var e4 = ErrAborted.WithMessage("custom message")
+	if e1 == e4 {
+		t.Fatal("should not equal the same error")
+	}
+
+	if !errors.Is(e4, e1) {
+		t.Fatal("errors.Is should be true, e1 is in the tree of e4")
+	}
+
+	if errors.Is(e1, e4) {
+		t.Fatal("errors.Is should be false, e1 is not a custom message")
+	}
+
+	if !errors.As(e4, &aborted) {
+		t.Fatal("errors.As should be true")
+	}
+
+	var custom customMessage
+	if !errors.As(e4, &custom) {
+		t.Fatal("errors.As should be true")
+	}
+	if custom.msg != "custom message" {
+		t.Fatalf("unexpected custom message: %q", custom.msg)
+	}
+	if custom.err != e1 {
+		t.Fatalf("unexpected custom message error: %v", custom.err)
+	}
```

Comment:
q: Was thinking of another case to compare using `errors.Is`.

But `errors.Is(e1, custom.err)` and `errors.Is(custom.err, e1)`. Both returns true. Is that an expected result?

Reply:
NOTE: It is true since either they are equivalent or wrap some equivalent error.